src/sonic/mod.rs

```diff
@@ -0,0 +1,534 @@
+use crate::kzg10;
+use crate::{BTreeMap, BTreeSet, ToString, Vec};
+use crate::{BatchLCProof, Evaluations, QuerySet, QuerySetError};
+use crate::{LabeledCommitment, LabeledPolynomial, LinearCombination};
+use crate::{PCRandomness, PCUniversalParams, Polynomial, PolynomialCommitment};
+
+use algebra_core::{AffineCurve, Field, One, PairingEngine, ProjectiveCurve, Zero, UniformRand};
+use core::marker::PhantomData;
+use rand_core::RngCore;
+use std::collections::HashMap;
+
+mod data_structures;
+pub use data_structures::*;
+
+mod error;
+pub use error::*;
+
+pub struct Sonic<E: PairingEngine> {
+    _engine: PhantomData<E>,
+}
+
+impl<E: PairingEngine> Sonic<E> {
+    fn accumulate_elems<'a>(
+        combined_comms: &mut HashMap<usize, E::G1Projective>,
+        g1_projective_elems: &mut Vec<E::G1Projective>,
+        g2_prepared_elems: &mut Vec<E::G2Prepared>,
+        vk: &VerifierKey<E>,
+        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Commitment<E>>>,
+        point: E::Fr,
+        values: impl IntoIterator<Item = E::Fr>,
+        proof: &kzg10::Proof<E>,
+        opening_challenge: E::Fr,
+        randomizer: Option<E::Fr>
+    ){
+        let mut curr_challenge = opening_challenge;
+
+        // Keeps track of running combination of values
+        let mut combined_values = E::Fr::one();
+
+        // Iterates through all of the commitments and accumulates common degree_bound elements in a HashMap
+        for (labeled_comm, value) in commitments.into_iter().zip(values) {
+            combined_values += &(value * &curr_challenge);
+
+            // Gets the degree_bound. If there is none, then use default bound of maximum degree
+            let degree_bound =
+                if let Some(degree_bound) = labeled_comm.degree_bound() {
+                    degree_bound
+                }else{
+                    vk.max_degree
+                };
+
+            let comm = labeled_comm.commitment();
+
+            // Applying opening challenge and randomness (used in batch_checking)
+            let comm_with_challenge: E::G1Projective = comm.comm.0.mul(curr_challenge);
+
+            if let Some(randomizer) = randomizer {
+                comm_with_challenge.mul(randomizer);
+            }
+
+            // Accumulate values in the HashMap
+            if let Some(combined_comm) = combined_comms.get_mut(&degree_bound){
+                *combined_comm += &comm_with_challenge;
+            }else{
+                combined_comms.insert(degree_bound, comm_with_challenge);
+            }
```

Comment:
This isn't urgent, but we can move to the [entry API](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry)